tests/test_run_reference_vmm.py

```diff
@@ -138,6 +139,43 @@ def expect_string(vmm_process, expected_string, timeout=TEST_TIMEOUT):
             raise
 
 
+def run_cmd_inside_vm(cmd, vmm_process, prompt, timeout=5):
+    """Runs a command inside the VM process and returns the output.
+
+       If the command runs successfully, output of the command is returned or else a
+       suitable exception is raised. The `timeout` parameter is used to indicate how
+       much time to wait for the command to complete.
+
+       Note: `cmd` and `prompt` should be a `bytes` object and not `str`.
+    """
+    cmd = cmd.strip() + b'\r\n'
+
+    vmm_process.stdin.write(cmd)
+    vmm_process.stdin.flush()
+
+    then = time.time()
+    giveup_after = timeout
+    while True:
+        try:
+            data = os.read(vmm_process.stdout.fileno(), 4096)
+            output_lines = data.split(b'\r\n')
+            last = output_lines[-1].strip()
+            if prompt in last:
+                # FIXME: WE get the prompt twice in the output at the end,
+                # So removing it. No idea why twice?
+                # First one is 'cmd'
```

Comment:
You can get it once because you write to stdin, and once because it is captured by the serial console I think.

Reply:
Well - I am not sure why it is obtained twice at the end :-( But this too will go away with `pexpect`.